integration/singularity/reader.go

```
@@ -0,0 +1,76 @@
+package singularity
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+
+	singularityclient "github.com/data-preservation-programs/singularity/client/swagger/http"
+	"github.com/data-preservation-programs/singularity/client/swagger/http/file"
+	"github.com/gotidy/ptr"
+)
+
+// io.ReadSeekCloser implementation that reads from remote singularity
+type SingularityReader struct {
+	client *singularityclient.SingularityAPI
+	fileID uint64
+	offset int64
+	size   int64
+}
+
+func (r *SingularityReader) Read(p []byte) (int, error) {
+	logger.Infof("buffer size: %v", len(p))
+
+	buf := bytes.NewBuffer(p)
+	buf.Reset()
+
+	if r.offset >= r.size {
```

Move checks earlier in the function to fail fast?